wacz-ipfs/latest/index.md

```diff
@@ -0,0 +1,134 @@
+
+## ZIP File Chunking
+
+The [ZIP file format](https://www.loc.gov/preservation/digital/formats/fdd/fdd000354.shtml) is used to create a single file which can contain several files within a directory structure. It allows files to be stored compressed and in their raw uncompressed form as continuous segments of the file.
+
+We can combine this with our custom chunking code in order to chunk a ZIP file at it's file boundries and generate regular UnixFS file nodes for the file chunks in the same way that they would be generated on their own.
+
+TODO: Diagram showing file boundries in a zip file, and chunk boundries in a UnixFS Dag
+
+## WARC File Chunking
+
+WARC files contain a series of "Records" for loaded resources.
+
+We group these records into individual UnixFS File DAGs so that they can more easily be recombined between archives at the UnixFS level.
+
+TODO: Diagram of file with deliniation for records. The 'type' of record can be seen and the request and response records are grouped together (outline with a box?)
+
+As well, we make sure to split out the response body into its own DAG node so that its contents may be deduplicated accross all archives and IPFS content.
+
+## WACZ File Chunking
+
+The WACZ file format is based on the ZIP file format with the addition of a WARC file for archival data, and some extra files for viewing the data.
+
+TODO: Diagram of directory structure within WACZ with the files usually added by WebRecorder tools
+
+We take advantage of this by reading the metadata from the WACZ and WARC files and creating additional chunking boundries at the boundries of individual files within the WARC file.
+
+TODO: Diagram of WACZ IPLD structure. Show ranges of zip files and the end directory structure, make boxes around IPLD boundries with labels.
+
+Due to how UnixFS file DAGs can be nested indefinately, we can build up the merkle DAG for the WARC file separately and embed it into the DAG of the ZIP file itself.
+
+Once a file has been chunked, any resources that have the same data within the archive or within other archives will get deduplicated at the storage / loading layer.
+
+## Algorithm
+
+This algorithm is used for generating a UnixFS file tree for a single WACZ file which contains a webarchive.
+
+This spec uses shorthands for referencing the UnixFS spec, and implementers should follow that spec in order to properly construct UnixFS Dag-PB blocks.
+Shorthands are also used when referring to the ZIP file structure such as parsing out headers or the final directory block.
+This spec relies on the existence of a blockstore to `put()` blocks of data into and get back the `cid` for the data.
+This could be an on-disk or in-memory repo, or a streaming CAR file generation library.
+
+- Create a root UnixFS.File entity representing the entire archive
```

Comment:
Hm, I wonder if this is the best way to explain how to do this. I was thinking more of having primitives like
`add()` and `concat()` and using these to explain how the format is created.
For example, 

```
Given primitives:
- add(bytes[]) : CID
- concat(CID 0, CID 1 ... CID N) : CID
ZIP = concat(add(local file header 1), add(file data 1), ... add(local file entry N), add(local file entry N), add(central dir record))

also referencing something like this diagram:
https://en.wikipedia.org/wiki/ZIP_(file_format)#/media/File:ZIP-64_Internal_Layout.svg

For WARC, we also define the parts, maybe showing a diagram like:
https://wiki.archivematica.org/File:WARCdiagram.png (or have our own with more detail) and then explain.

WARC Record = WARC Header + WARC Payload + \r\n
WARC Response Record + WARC Request Record = WARC Response Header + WARC Response Payload + \r\n + WARC Request Header + WARC Request Payload + \r\n

One way to group this for IPFS, what we're currently doing:

WARC Record Group: CID = concat(add(WARC Response Header), add(WARC Response Payload), add(\r\n + WARC Request Header + WARC Request Payload + \r\n))
WARC Record Groups = concat(WARC Record Group 1, ... WARC Record Group N)

Perhaps this should go into the top-level description above?